llvm/lib/MCCAS/MCCASObjectV1.cpp

```diff
@@ -1905,7 +1906,14 @@ struct DIEDataWriter : public DataWriter {
 /// is described by some DIEAbbrevRef block.
 struct DistinctDataWriter : public DataWriter {
   Expected<DIEDistinctDataRef> getCASNode(MCCASBuilder &CASBuilder) {
-    return DIEDistinctDataRef::create(CASBuilder, toStringRef(Data));
+    SmallVector<uint8_t> CompressedBuff;
+    compression::zlib::compress(arrayRefFromStringRef(toStringRef(Data)),
+                                CompressedBuff);
+    SmallVector<char> FinalBuff;
+    raw_svector_ostream OS(FinalBuff);
+    encodeULEB128(Data.size(), OS);
+    FinalBuff.append(CompressedBuff.begin(), CompressedBuff.end());
```

Comment:
Are we doing this copy just so we can prepend the size of the uncompressed data?
Here's a suggestion instead, let's use the "Pad To" argument of `encodeULEB128` and put the size at the end, always with 8 bytes.

```
     SmallVector<uint8_t> CompressedBuff;
     compression::zlib::compress(arrayRefFromStringRef(toStringRef(Data)),
                                 CompressedBuff);
     CompressedBuff.append(8, 0);
     encodeULEB128(Data.size(), CompressedBuff.end() - 8, 8 /*Pad to*/);
     FinalBuff.append(CompressedBuff.begin(), CompressedBuff.end());
```

The decoder would decode using the last 8 bytes, and decompress using [start, end-8]